demos-europe/demosplan-core
# feature (refs T29261): handle addon permission

**Ticket:** https://yaits.demos-deutschland.de/T29261

**Content:**

* adds the `ResolvablePermission` class, which is responsible to carry permission settings ("should a permission be enabled or not?") and will at some point in the future stored in the DB as entity (but not now)
* adds the `PermissionResolver` service, which is able to determine if a permission should be enabled or not based on the settings in a `ResolvablePermission` instance
* moves the code to load the `permissions.yml` from the `Permissions::setInitialPermissions` into a separate `CachingYamlPermissionCollection` service, so it can be reused by addons that want to utilize it
* adds support for the `description` field to the `Permission` class, so it is loaded from the `permissions.yml` like the label already is

These are the base changes of the permission system needed in the core in the future. Other parts, like classes to easily create `ResolvablePermission` instances will implemented outside of the core. Hence of the added classes only `CachingYamlPermissionCollection` is in use.

**Background:**

This is the initial concept to allow addons to the dplan permission system. Assumptions and reasoning:

* Addons can provide the core with defaults for decisions if a permission should be enabled or not, but at some point in the future those need to be adjustable by the platform administrator via UI to satisfy different requirements of different platforms. Ignoring this requirement for now and refactoring all addon permissions later didn't seem to be efficient.
  * To allow the adjustments via UI the decisions need to be storable in the database, which will be possible with the Drupal filter format. When an addon provides the core with its defaults in the Drupal filter format, adjustments of these defaults with pretty UI elements become possible too.
* addons can only enable permissions, not disable them
  * this deliberate limitation is the result from the learning that disabling permissions introduces too much complexity
* addons should not be able to check for permissions defined by other addons, as if addons would do that they'd become to fragile
  * for now in this PR this is done by putting the addon scope information into the permission name, which feels like an ugly comeback of the filestring approach
  * Allowing an addon B to check permissions of another addon A that is defined as dependency of addon B will be implemented in the future. If we can use PHP enums at that point it may become very easy and a replacement for the current permission name based scoping.
* the creation of permission decisions (i.e. in this PR in `MaillaneAddonActivator`) should be as readable as possible. In this PR this was attempted by providing builder methods to create the Drupal filter with purposely little abstraction, but I'm not convinced by the result as it still seems too complex, especially when used with `AND`/`OR` groups. A `ConditionFactory` or `FluentQuery` approach may provide better results.
* The format used in this PR is not exactly the Drupal filter format. We needed an additional boolean `parameter` field. This field determines if the `value` field in a Drupal filter condition is actually meant for comparison with the value stored in the object or if the `value` is a reference to a variable of the current context (i.e. procedure, customer or user). The example permission used in this PR shows the necessity to give access to the context. I'm not yet happy with this approach. Beside the awful naming it seems a bit hacky. Maybe it can be abstracted away as the use case it is needed for here (roles in a specific customer) applies to almost every permission.

The `DplanAddon` and corresponding repository are of no relevance for this PR.

### How to review/test

The important parts of the approach are implemented, but except for the `CachingYamlPermissionCollection` they can not be tested in the application yet. If you want to understand the implementation I suggest to review in PHPStorm and try to understand the implementation by reading the docblocks and comparing it to the background described above.

As my local environment is currently broken I would be really glad if you could test this locally on your machine on any project to see if the `CachingYamlPermissionCollection` is working as intended.

### Linked PRs (optional)


### Tasks (optional)


### PR Checklist


Delete the checkbox if it doesn't apply/isn't necessary.
